### ISSUE TITLE
GitLab integration

### ISSUE THREAD

**Issue by tommoor** [outline/outline#6795]
After https://github.com/outline/outline/issues/6406 similar to #4427

**Comment by almereyda** [comment]
To note, there is a basic GitLab integration _snippets_ from the hosted instance at `gitlab.com`, which was contributed with #6217, #6290 and https://github.com/outline/outline/commit/8b6a760e02a5a30d4f08db9e34682fb4d6840f0a

The use case for self-hosted snippets is documented in:

- https://github.com/outline/outline/discussions/6427

with some recent activity in the linked

- #6317

that leads us to

- https://github.com/IamTaoChen/outline/tree/gitlab_new

**Comment by thedarkside** [comment]
That is what we are waiting for! 

I would like to outline our requirements for a good integration here. 

Besides the ability to embed issues, epics and merge requests from a selfhosted gitlab instance it would be beneficial when the embeds **update automatically** showing the current status like if an issue is open or closed, when was the last change etc. 

Because our instance is private only, outline would need an api key to fetch the information.

Is this kind of integration planned?

**Comment by almereyda** [comment]
#6406 has landed with #6414, which could provide as a nice blue print for the API surface of the integration to cover on Outline's side. It should be not too hard to find equivalents for the API surface used on GitHub's side in GitLab.

**Comment by JoaquinLop** [comment]
It would be great to have this functionality.

**Comment by rausub** [comment]
i concur it would be nice to have this

**Comment by almereyda** [comment]
@tommoor re: https://github.com/outline/outline/issues/6407#issuecomment-2652602393

In case one wanted to continue here from the community-side of things, should we also wait for this "solution to be pluggable" or might GitLab rather just be the next work item bringing with it the decoupling mentioned in the linked comment, also paving the way for Linear and sorts?

**Comment by tommoor** [comment]
> should we also wait for this "solution to be pluggable"

Yep, not sure on the priority of GitLab support itself for the maintainer team, it's not requested all that often

**Comment by almereyda** [comment]
For us it is a priority and I would put in significant effort into porting any existing integration to GitLab, which is mostly copy and paste and reading the documentation about API endpoints.

Will return here when the "pluggable solution" exists and can be adapted from.

**Comment by almereyda** [comment]
With #6406 and #6407 having landed in v0.84.0, the required prior art for an adaptation to GitLab _and potentially other systems_ #6507 has arrived.

**Comment by alphanso** [comment]
We are also evaluating outline. We are using gitlab self-hosted along with slack and Google workspace. Self-hosted Gitlab integration will certainly help the case. Thanks

**Comment by Mtze** [comment]
I would also be very interested in this feature. :)

**Comment by fmotzet** [comment]
We are also evaluating Outline, coming from Confluence and this feature would be amazing for us.

**Comment by tommoor** [comment]
Should tackle this soon

**Comment by EHRETic** [comment]
Would also be interested with this (Gitea for me)! 👌😉

**Comment by CanalWood** [comment]
Same for me, a Gitlab integration will be great.
It will be great to have a my page as markdown format stored in Gitlab and the render into Outline.

**Comment by almereyda** [comment]
> It will be great to have a my page as markdown format stored in Gitlab and the render into Outline.

That's maybe a separate issue/feature request. Look at the existing integrations for Linear and GitHub. This is to show previews for issues, MRs etc. in private projects.

> The [GitHub integration](https://docs.getoutline.com/s/guide/doc/github-fSO93uaeYI) allows previews of issues and pull requests in GitHub within Outline. For example, hovering over a GitHub url will show a preview of the issue status at that url.
>
> https://docs.getoutline.com/s/hosting/doc/github-GchT3NNxI9

The linked guide has more details about how this works.

- [GitHub - Outline](https://docs.getoutline.com/s/guide/doc/github-fSO93uaeYI)
- [Linear - Outline](https://docs.getoutline.com/s/guide/doc/linear-nXLg6eskUZ)

**Comment by dizummd21** [comment]
Yes GitLab Integration where each user can add tokens to display tickets of self hosted gitlab projects in different Gitlab instances..

**Comment by tommoor** [comment]
This is now live in cloud-hosted Outline and will be available in self-hosted `v1.6.0` in March

**Comment by ashleysommer** [comment]
@tommoor can you confirm please, does this feature (GitLab integration, now live) also support Gitea and Forgejo as suggested in the thread above?

**Comment by tommoor** [comment]
Nope, those are separate platforms – just GitLab

**Comment by almereyda** [comment]
A first feedback from using this, which looks excellent. Looking forward to full card previews ✨ .

---

For those who were wondering how to configure the new OAuth application on the GitLab end, the scopes used are:

- **read_api**
- **read_user**

https://github.com/outline/outline/blob/12c71f267e7e598b8b5fe47e5d70fce0ce6c90ef/plugins/gitlab/shared/GitLabUtils.ts#L85

Both could be mentioned on a future documentation page under

- [Integrations - Outline](https://docs.getoutline.com/s/hosting/doc/integrations-Oav5MXNDJk)

which in return could be linked from the Connect modal, like with (a small, grey question mark bubble and) a "Help" text right after "instance".

<img width="976" height="620" alt="Image" src="https://github.com/user-attachments/assets/6f04d440-073b-40ec-a1dc-426125804646" />

---

To make the onboarding flow for user's even simpler ("What is an OAuth Client Application?", "Where do I need to create it?"), Dynamic Client Registration could be used. Then one only needs to enter the GitLab URL and can be done after a confirmation flow.

- [Final: OpenID Connect Dynamic Client Registration 1.0 incorporating errata set 2](https://openid.net/specs/openid-connect-registration-1_0.html)
- [RFC 7591 - OAuth 2.0 Dynamic Client Registration Protocol](https://datatracker.ietf.org/doc/html/rfc7591)
- [Use OpenID Connect as an authentication provider | GitLab Docs](https://docs.gitlab.com/administration/auth/oidc/)
- [Implement OAuth 2.0 Dynamic Client Registration Protocol (#554820) · Issue · gitlab-org/gitlab](https://gitlab.com/gitlab-org/gitlab/-/work_items/554820)

This is also used for MCP clients and already implemented there.

---

Should I file separate issues for both, GitLab integration documentation + Dynamic Client Registration?

If not used, yet, the latter probably also exists for Linear and GitHub.

**Comment by T-Herrmann-WI** [comment]
Dear @tommoor , it looks good. It would be very helpful to show also ticket lists instead of only one ticket of a project. Best Tim
